### DISCUSSION TITLE
Document API in Stargate V2

### DISCUSSION THREAD

**Discussion by ivansenic** [stargate/stargate#1742] [RFCs]
This discussion presents a general overview on the possible design of the Document API in Stargate V2. It presents different solutions, opens different questions that should be answered and provides a basic overview on how the execution can be planned and estimated.

# Introduction

## Stargate V2

The Stargate version 2 project has been planned during the 2021 and started in the Q4 2021. The main goal of this project is to divide the Stargate monolith based on OSGi into a more modern microservice-based architecture. There are different reasons why the Stargate team has decided to go this path, and this has been documented already in several places. Thus, it’s a good idea to read available resources on Stargate V2 before taking on this document:

* [Stargate V2 Design](https://github.com/stargate/stargate/discussions/1381)
* [Stargate V2 / REST / Schema access optimization: partial caching (f.ex of PKs)](https://github.com/stargate/stargate/discussions/1481)
* [Stargate Code Refactoring (V1->V2)](https://github.com/stargate/stargate/discussions/1454)
* [Select framework for Java-based API Service implementations ](https://github.com/stargate/stargate/discussions/1526)

## How’s Document API different from other Stargate APIs

It’s also important to point out that the Document API is different from other APIs in the Stargate. We like to say that other APIs are CQL-first and that Document API is not. The CQL-first APIs are mapping each operation to a single operation against the underlying Cassandra. Thus, in a way, they expose Cassandra functionality over REST, gRPC or GraphQL.

The Document API adds additional functionality, storing the JSON documents, which is not directly supported by Cassandra. In order to do this, operations against Document API can not be directly mapped to Cassandra, and usually each operation has to perform multiple statements against it in order to complete successfully.

The Document API stores a JSON document in multiple rows, using the JSON shredding technique, where each row stored will represent a single value of the JSON document. Thus, the number of rows for a single document, depends on the document size.  outlines why this model was chosen.

The Document API also exposes additional filter operations that are not directly supported by Cassandra. For example, it provides support for _$and_, _$or_ and _$not_ expression, adds filters like _$nin_ (not in) and _$exists_, etc.

# Design Possibilities

## Main Challenges

There are two main challenges in the Document API that heavily influence the possible design of the service in the Stargate V2.

1. **Amount of statements for a single operation.** As already mentioned, amount of C* statements executed for a single Document API operation vary. For example, in order to search documents, the API will execute at least one statement for each expression provided, and then an additional statement per document population. The network overhead, thus, becomes a growing issue, as the amount of back and forth operations to be executed against the gRPC bridge is not constant. In addition, as the authorization over gRPC bridge is executed for each statement, we would pay additional performance penalties for document writes and reads, as instead of doing the authorization once per Document API operation, we would do it for each statement executed.
2. **Documents ending in the middle of the Cassandra page.** As documents are stored in multiple rows, a single document can end anywhere in a Cassandra page. This means that the page state of the Document API is not the same as the page state in Cassandra. We have developed a solution in Stargate to “break” a Cassandra page in the middle, so that we can correctly mark the end of a document and enable a user to continue document fetching from that same position in a follow-up query.

These challenges make it very hard for the Document API to operate against the currently exposed public gRPC service that supports plain-text queries and batches. This is why we started to think about a new, internal, gRPC API that would expose a better fitting API for operating with documents, or building on top of the gRPC bridge that will be introduced for Stargate V2.

## Proposals

The three proposals presented in this document are focused around introducing new functionality in the internal gRPC based API. The main difference between them is how they are designed and what kind of operations they expose over gRPC. However, effectively they follow the Stargate V2 design approach:

![image](https://user-images.githubusercontent.com/10600041/160610340-82cd584a-7bf6-45b7-bb4e-79f73f843ddc.png)

### Proposal I: Higher level API

Higher level gRPC API would represent a generic abstraction for handling document related operations. This API would be dedicated to serve the Documents API. The API would be generic in a way that there could easily be other implementations that don’t store documents in Cassandra, but in some other document database or storage. Thus, it would be completely agnostic from Cassandra.

With this approach we would highly favor the open data gateway, as we would expose possibilities for open source contributions in Stargate that don’t target C*.

Designing this API should be in a way of taking best practices in the existing document databases. Proper analysis of MongoDB, but other document DBs as well, could provide us with a generic way to support different features, both the ones that exist currently in Stargate, but also ones that will be added later in the future. This would also be a good chance to perform a deeper comparison and analysis of missing document features in Stargate.

A quick check showed that this API could be implemented as the wrapper around two existing document services, [DocumentWriteService](https://github.com/stargate/stargate/blob/master/restapi/src/main/java/io/stargate/web/docsapi/service/write/DocumentWriteService.java) and [DocumentSearchService](https://github.com/stargate/stargate/blob/master/restapi/src/main/java/io/stargate/web/docsapi/service/query/DocumentSearchService.java). In order to support current features, the minimum API would need to support following commands (operations):

* Insert
* Update
* Patch
* Delete
* Read
* Search

Additional points about the API design are considered and discussed in the [gRPC API Design](#grpc-api-design) and [Authentication & Authorization](#authentication--authorization) sections. 

#### Pros & Cons

<table>
   <tr>
      <td><strong>Pros</strong>
      </td>
      <td><strong>Cons</strong>
      </td>
   </tr>
   <tr>
      <td valign="top">
         <ul>
            <li>Does not suffer from any problems or potential performance penalties described in the <a href="#main-challenges">Main Challenges</a>.
            <li>Arguably easier to add features in the future (aggregation, sorting, etc.) as we are closer to the persistence.
            <li>Opens Stargate as the open-source data gateway (and not only the open-source Cassandra gateway).</li>
         </ul>
      </td>
      <td valign="top">
         <ul>
            <li>Leaves a solid amount of the code in the Coordinator for the actual API implementation. It’s not clear if this is breaking the original motivators for the Stargate V2.</li>
            <li>Introduce tight coupling between Document API and the Bridge, which makes it hard to independently manage and deploy components.</li>
         </ul>
      </td>
   </tr>
</table>

### Proposal II: Lower level API

Lower level API proposal should represent addition to the V2 gRPC bridge in order to close the existing gaps that prevent usage of the publicly available gRPC endpoint for the Document API. It’s a minimal API that would actually ensure that most of the logic currently being implemented for the Documents API in the coordinator is moved to the REST side. However, it also implies that additional network hops would be added, as well as authentication on the Statement level.

This API is agnostic of documents and is tightly coupled to Cassandra. In practice, this should be a wrapper around the [QueryExecutor](https://github.com/stargate/stargate/blob/773eacd0056de88defdb3944c4dcb3d4b07a01fb/restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java) service that enables different features for the document querying (page state at the end of the document, merging multiple queries, etc.). Although this class is docs-specific, it might be possible to refactor it in a way that has not direct relation to docs, but rather collection of C* rows. This might open new possibilities for the other APIs as well, as it could provide a way for them to utilize some features here, like the merging of multiple queries.

However, seems that the refactoring needed to be done here is quite huge:

* Making [QueryExecutor](https://github.com/stargate/stargate/blob/773eacd0056de88defdb3944c4dcb3d4b07a01fb/restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java) more abstract and document agnostic
* Making [QueryExecutor](https://github.com/stargate/stargate/blob/773eacd0056de88defdb3944c4dcb3d4b07a01fb/restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java) to work with gRPC queries and not the current model around query builder
* Refactoring all the components depends on the [QueryExecutor](https://github.com/stargate/stargate/blob/773eacd0056de88defdb3944c4dcb3d4b07a01fb/restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java) to switch to gRPC query building (CQL + separate bindings). Although this is not hard, adapting tests will be a big task. All document unit tests depend on the ValidatingDataStore which would be unavailable now.

#### Pros & Cons

<table>
   <tr>
      <td><strong>Pros</strong>
      </td>
      <td><strong>Cons</strong>
      </td>
   </tr>
   <tr>
      <td valign="top">
         <ul>
            <li>Moves most of the code and processing from coordinator to the REST side. Coordinator would not know anything about the documents in fact.</li>
         </ul>
      </td>
      <td valign="top">
         <ul>
            <li>Should suffer from performance penalties described in the <a href="#main-challenges">Main Challenges</a>. Increased network hops and more auth calls (batching authz can be solved internally in Stargate, see <a href="https://github.com/stargate/stargate/issues/1587">stargate#1587</a>).</li> 
            <li>Amount of work needed to be done for refactoring and test fixing seems high.</li>
         </ul>
      </td>
   </tr>
</table>

#### Investigation Considerations (2022-02-25)

After attempting to implement just the Documents API’s **QueryExecutor**, a few challenges have arisen that are worth mentioning and discussing:

* At its core, QueryExecutor uses an implementation of `DataStore` to do all of its query handling and to get more internal information about complex paging states, etc. The gRPC implementation is not based on a DataStore and instead directly uses the Connection object.
* QueryOuterClass.Query objects have no information on hand about the table (or columns) that they are querying; this poses a challenge for the QueryExecutor because it collates and sorts results based on the “path” column values on the final result set.
* With that in mind, the ResultSet that comes out from gRPC also does not have anything indexed by column name (and is instead indexed positionally to access particular values)
* There are likely other objects that need adapting that QueryExecutor uses.
* Therefore, splitting the QueryExecutor in a manner where each of its query executions is instead a plain CQL query over gRPC is possible (Proposal II), but also very tedious and potentially requires changing the above pieces in a way that won’t match gRPC implementations for the other APIs. I decided to stop breaking it apart further when I realized how tedious it would be as a POC, and instead come back here to discuss.
* The other option would be putting the entire QueryExecutor behind gRPC, living in the `grpc` module as-is, and then all gRPC operations against it would be created specifically for the Documents API. (this is Proposal I). This is certainly less tedious because the meat of the execution logic (QueryExecutor) remains as-is, without much need for change. It would still be backed by a DataStore object, and would use the objects that it already is built to use (BuiltSelect, Row, ResultSet, etc)

Something that has not yet been discussed but might be a useful point: If we went the route of Proposal I, users would also have the option of potentially using the Documents API directly over gRPC. While it seems trivial and maybe even less ergonomic than using HTTP, it actually could be useful for integrations such as the Mongoose integration (cutting out extra middle-men like HTTP might be useful), or for users who exclusively use gRPC and want to choose their own entry point.

### Proposal III: Simple bridge extension

The third proposal came to our intention after the initial discussions focused on the first two proposals. The idea is simple, try to move the complete Document API source code to the new microservice and provide the bare minimum extension to the bridge to support this. This represents even the lower gRPC API than proposed in the [Proposal II](#proposal-ii-lower-level-api), as QueryExecutor would be on the client side.

There are two mandatory requirements for this proposal to be possible:

* Be able to transfer row-based page state to the client.
* Be able to transfer a comparable object for each row over gRPC, so that rows can be compared for the partition order on the client side.

The analysis outlined in the [QueryExecutor on the client side (analysis)](#queryexecutor-on-the-client-side-analysis) showcased that these two requirements can be met.

#### Pros & Cons

<table>
   <tr>
      <td><strong>Pros</strong>
      </td>
      <td><strong>Cons</strong>
      </td>
   </tr>
   <tr>
      <td valign="top">
         <ul>
            <li>Since QueryExecutor is also on the client side, it’s expected that there will be less work than Proposal II. It’s only necessary to exchange interactions against DataStore to Bridge. 
            <li>Moves most of the code and processing from coordinator to the REST side. Coordinator would not know anything about the documents in fact.
            <li>Completely aligned with the Stargate V2 architecture.</li>
         </ul>
      </td>
      <td valign="top">
         <ul>
            <li>Should suffer from performance penalties described in the <a href="#main-challenges">Main Challenges</a>. Increased network hops and more auth calls (batching authz can be solved internally in Stargate, see <a href="https://github.com/stargate/stargate/issues/1587">stargate#1587</a>).</li>
         </ul>
      </td>
   </tr>
</table>

#### QueryExecutor on the client side (analysis)

QueryExecutor depends on two methods from the Persistence implementation that are needed to be implemented over gRPC bridge in order to be able to move QueryExecutor to the client side. These are:

* io.stargate.db.Persistence.Connection#makePagingState
* io.stargate.db.Persistence.Connection#makeRowDecorator

This means that results of these methods should be transferable to the client side.

Make paging state is possible to transfer as at the end it’s just a plain ByteBuffer. However, the _RowDecorator_ implementation seems tightly coupled to the Cassandra types. Effectively, the decorator is only needed so that we can construct the _ComparableKey_ that’s used to sort documents based on the partition position. The sorting is delegated to the _DecoratedKey_, which is using a token and key for doing the comparison.

It’s unclear how we can construct such a comparator on a client side. Effectively we only need that comparator.

The _DecoratedKey_ in DSE has a method _byteComparableAsString_. The string that's the result of this method can be safely used as a comparison. Quick tests showed that switching to this string as a comparator works fine and that our tests are green.

However, C* 3.11 and C* 4 do not expose such a method in the _DecoratedKey_. Maybe there is a possibility that we can construct that on our own, but it seems that would be a major hack. The issue for adding this exists for over 8 years now ([CASSANDRA-6936](https://issues.apache.org/jira/projects/CASSANDRA/issues/CASSANDRA-6936)), with the PR in review since October 2021 ([apache/cassandra#1294](https://github.com/apache/cassandra/pull/1294)). This could mean that:

* Soon existing methods will be existing in C*
* If the "soon" takes too much time for us, we can always relate to that PR in order to temporarily copy the needed code to construct the bytes comparable directly in Stargate persistence modules.

## gRPC API Design

This section opens some design questions and tries to identify differences between two proposals above.

#### JSON-aware or not

It’s clear that [Proposal II: Lower level API](#proposal-ii-lower-level-api) and [Proposal III: Simple bridge extension](#proposal-iii-simple-bridge-extension) will not be aware of the JSON. It will be operating directly with the queries and rows.

For the [Proposal I: Higher level API](#proposal-i-higher-level-api), we could go either way. For example, we can say that the new gRPC Document API receives the minified JSON as the _byte[]_ array and responds with the same. On the other hand, we could completely substitute the JSON strings and communicate with our own data structure resulting from JSON shredding. Meaning, we would be able to send back and forth document rows, where each row represents one JSON value in the tree. This would also imply that other places where JSON is used (like _where_ clause, _fields_ array) are having a well-defined model.

It’s hard to say what is better. Doing shredding on the REST side does offload some processing from a coordinator. On the other hand, for deeply nested documents, transferring rows over the network would require more bytes than minified JSON string.

#### Cassandra aware or not

Again, it’s clear that [Proposal II: Lower level API](#proposal-ii-lower-level-api) and [Proposal III: Simple bridge extension](#proposal-iii-simple-bridge-extension) will be aware of Cassandra.

The [Proposal I: Higher level API](#proposal-i-higher-level-api) should not be aware of Cassandra. All Cassandra level properties (like max depth of Document table, consistency level of queries, etc.) should be defined on the coordinator side and the REST should not even be aware of it. However, this brings a question about schema awareness. If the REST side is not aware of Cassandra we can not be aware of the schema as well. Other APIs will be aware of the schema, which will enable them to execute certain tasks on the REST side, here for example, checking if a table exists and is a valid document table. Meaning, that these responsibilities would be taken care of in the coordinator itself, again pushing more responsibility to that side.

#### Batches

Current support for batch operations in Stargate Document API is limited to batch inserts. It’s reasonable to assume that this is not enough and that batching should be possible for all write operations.

This opens questions about the proper batching support in the future. MongoDB, for example, has an _[ordered](https://docs.mongodb.com/manual/reference/method/db.collection.insert/#insert-multiple-documents)_ property for batch inserts, where documents are inserted in a given order and if a document insertion fails, it stops further inserts. We might add similar features in the future.

The difference between proposals is where the control of such execution flow is taken care of. In the case of the higher level API, the control is on the coordinator, while in the other proposals it should be on the REST side. It’s again some kind of tradeoff, less gRPC requests and more work on the coordinator for the first proposal, and vice versa for the other ones.

#### Future work (TTL, sort, aggregations)

Here is a quick overview on the possible future work and the difference between proposals:

* **Document Time-To-Live**. Not having many differences. Except for patches where if TTL is defined we would need to do a read first, which means one extra network call in the proposal two.
* **Sorting**. If based on SAI indexes, as planned now, there should be no major differences nor different implementation estimates.
* **Aggregations**. This is not an easy one, as we are still not aware of what kind of aggregations we want to support. In the case of multi-stage aggregations supported by MongoDB, it seems that a higher level API would be beneficial here.

## Authentication & Authorization

The authentication and authorization is an important aspect and should be clearly defined how it would be implemented in the two proposals.

For the [Proposal I: Higher level API](#proposal-i-higher-level-api), this would stay more or less as is at the moment. The REST side would transfer the_ X-Cassandra-Token_ received in the gRPC call context. From there on, based on the operation being executed we would authenticate and authorize the operation. Considering that authentication details could be cached for the token, we would be seeing one authorization call per op. This should provide better performance because it does the authorization per document operation on the coordinator.

For the [Proposal II: Lower level API](#proposal-ii-lower-level-api) and [Proposal III: Simple bridge extension](#proposal-iii-simple-bridge-extension) we would be using the existing concepts from the publicly available gRPC. This is what REST API V2 is doing at the moment. The difference from the above, would be that we can not authorize per document operation, but rather per Statement executed. It was already mentioned several times that this can lead to certain performance penalties.

#### Auth on REST

Note that for the all solutions there is a possibility to offload the authn/authz to the REST. The idea is that the REST side authorizes on the document operation and then executes a gRPC call via internal bridge that requires no auth. This is very similar to the work that did for the CQL bridge, but just for the gRPC. This can result in multiple benefits:

* No auth needed on the coordinator.
* Always does authz on the Document operation and not the particular Statements.
* Fail fast in case of the wrong auth, without even reaching the coordinator.
* Plug-in any implementation on the REST side, meaning developers can use standardized authn/authz solutions that are not defined explicitly by Stargate (this is even possible to migrate to other APIs).

In order for this solution, we would need to actually properly define roles on the REST side that are not coupled to the Stargate/Cassandra. Then we need the authn/authz implementation that bridges over to the Stargate Auth API. For there on we can use any existing implementation (table-based, Astra, etc).

## REST Service Responsibility

The design should clearly define what should be the responsibility for the REST Service. These responsibilities should be more or less shared with other Stargate REST APIs (REST, GraphQL).

* **Validated configuration.** Try to configure everything in a proper validated configuration model. Provide extensive documentation for each property. Provide meaningful development and production profiles.
* **Well-defined and documented HTTP API.** Properly implement the OpenAPI v3 and Swagger (document all possible responses, provide better DTO model and examples, etc).
* **Validation of the input.** Fail fast on invalid input with proper messages.
* **DTO transformation to gRPC model and vice-versa.**
* **Unified exception handling.** At best, this should be shared across APIs.
* **Retries**. Adds resilience on failing gRPC or data storage calls. A new feature we did not have so far.
* **Handle raw/not-raw document requests.** Document-internal feature.
* **JSON Shredding.** Only if a non-JSON-aware solution is selected.
* **Schema validation.** Only if Cassandra-aware.
* **Tenant-awareness**. Implement support for multi-tenancy, which would enable hooking in a tenant resolver and delegating tenant id in the gRPC context.
* **Metrics**. Expose HTTP server request and gRPC client request stats.
* **Tracing**. Basic support for the tracing of outgoing service calls.
* **Proper service integration tests.** Go away from doing all the integration tests against a running coordinator & data store. Focus on correctly utilizing the gRPC APIs only.
* **Contract testing.** Make sure that the OpenAPI specifications do match the API current state.
* **End-to-end testing**. Create a valid strategy for correct end-to-end testing.

# Final Design

This section outlines the final design chosen for the Stargate Document API V2. It’s based on the [Proposal III: Simple bridge extension](#proposal-iii-simple-bridge-extension), and the picture below depicts the main components that will be further discussed.

![Document API Microservice](https://user-images.githubusercontent.com/10600041/160610492-06e7f16c-4751-47be-ae3e-783e908747d3.png)

### Bridge extension

The required bridge extension needs to implement one additional method. This method will accept the existing gRPC-based Query as the input, but return the extended Response. This response will not only contain Rows, but for each Row it will contain the page state for that row and the comparable bytes that can determine the partition order of that Row.

These changes should be relatively simple, but do depend on the [apache/cassandra#1294](https://github.com/apache/cassandra/pull/1294) for supporting Cassandra 3.11 and 4.0. See [Risks](#risks) for further discussions on possible risk and available mitigations.

### Table & Keyspace Manager

Table and Keyspace managers would replace the current DocumentDB and take responsibility for creating and deleting document based tables & keyspaces. In addition, the Table manager is responsible to ensure that the target table is a valid Document API table. It will depend on the schema information provided by the _Stargate V2 Common_ project.

Note that the Table manager needs to ensure the existence of the valid target document table for every operation. Because of that, it’s very important that this component is performant. It’s expected that some kind of extra caching would be added here, so that we have minimum latency overhead.

### Document Service

Document service has the responsibility to orchestrate document operations:

* Shred documents
* Parse conditions and expressions
* Delegate reads and writes to the read/write bridge services
* Manage TTL
* Orchestrate batch document writes

It should stay almost the same as the current _ReactiveDocumentService_ class. We might decide to split it into a read and write service in order to decrease the size and the complexity of the current class.

### Write/Read Bridge Service

Write and Read Bridge services are responsible for creating the needed queries and executing those over the bridge. In case of the Read service, the execution is delegated to the Query Executor that will communicate with the extended bridge service.

These services should be a copy of the existing _SearchDocumentService_ and _WriteDocumentService_, with one important change. All the existing queries must be translated to the gRPC-based model, from the existing `persistence-api` model. This will be one of the major tasks in the complete V2 project, especially with respect to the testing, where numerous unit tests have to be rewritten, so that _ValidatingDataStore_ is replaced with the mocked bridge implementation.

### Query Executor

Should stay the same as now, just using the _Bridge_, instead of the `persistence-api` _DataStore_.

### Utilities

All the service utilities (JSON shredding, Expression parser, etc.) that we use right now will need only minor changes.

### HTTP Controllers

The HTTP controller will have some responsibilities mentioned in the [REST Service Responsibility](#rest-service-responsibility):

* Well-defined and documented HTTP API
* Validation of the input
* Unified exception handling
* Retries
* Handling raw/not-raw document requests
* Tenant-awareness
* Token-existence
* Observability (metrics, tracing)

## Authentication and authorization

The authentication and authorization will be done on the coordinator side. We will take the same approach currently implemented in the REST API V2 and GraphQL API V2, that simply passes the authentication token as the header when executing the gRPC call.

This document already outlined that this can represent some performance overhead, as now each query has to be authorized. Until now, we did authn/authz in the APIs directly interacting with the _AuthenticationService_ and _AuthorizationService_. This resulted in the single authn and authz call per document operation.

The good news is that existing auth implementations do implement short-lived caches. These caches save unnecessary repeated calls to Cassandra or external auth service when checking the same policy in the short time-frames.

The option to implement authn/authz on the REST side is still a viable option and possible upgrade in the future.

## Multi-tenancy

Multi-tenancy is one of the topics that should be addressed separately, as at the moment it’s not clear how the multi-tenancy should be implemented in the Stargate V2. From the external point of view, these should be satisfied:

* Multi-tenancy is optional, meaning that Stargate V2 can be used with a single tenant.
* Multi-tenancy is an explicitly available option, meaning that we should not hide that open-source Stargate V2 can be used as a multi-tenant application. This means that support for multi-tenancy is there and can be adapted to user needs.

## Framework

We are dedicated to moving to a new framework from day one of the Document API V2 project. We see no reason to stay on Dropwizard, even if this could provide us with a faster execution time due to the possibility to do more copy & paste. We believe that many out-of-the box solutions provided by the modern microservice frameworks will pay off in the long run, and will save us from a lot of unnecessary tasks.

In the [Select framework for Java-based API Service implementations](https://github.com/stargate/stargate/discussions/1526) we compared three dominant Java microservice frameworks. Although unofficial, it seems that Quarkus has won, and we are happy to bootstrap the project with it.

## Risks

The table below outlines few risks that have been identified for the Document API V2 project:

<table>
   <tr>
      <td><strong>Name</strong>
      </td>
      <td><strong>Description</strong>
      </td>
      <td><strong>Risk Level</strong>
      </td>
      <td><strong>Mitigation</strong>
      </td>
   </tr>
   <tr>
      <td valign="top">Byte comparable not available
      </td>
      <td valign="top">The C 3.11 and 4.0 don’t merge the bytes comparable PR until the date we want to release the new microservice.
      </td>
      <td valign="top">Medium
      </td>
      <td valign="top">No mitigation. In case it happens, copy the required code from the available PR and add to the Stargate persistence modules directly.
      </td>
   </tr>
   <tr>
      <td valign="top">Performance penalties
      </td>
      <td valign="top">Performance penalties due to the multiple network hops, schema auths, schema fetches, authorization per query, etc.
      </td>
      <td valign="top">Medium / High
      </td>
      <td valign="top">Implement and test the document insertion early in the development phase. Compare performance using already existing NoSQL bench scripts locally.
      </td>
   </tr>
   <tr>
      <td valign="top">Multi-tenancy
      </td>
      <td valign="top">Still unclear what the final requirements for the multi-tenancy are. Unclear how this will influence the interaction with the CNDB.
      </td>
      <td valign="top">Medium
      </td>
      <td valign="top">Implement components with multi-tenancy in mind, try to support all possible cases. Try to raise concern and find decisions in the team. 
      </td>
   </tr>
   <tr>
      <td valign="top">New framework
      </td>
      <td valign="top">Learning curve for the new framework could be steep and slow down the project.
      </td>
      <td valign="top">Low
      </td>
      <td valign="top">Find all available resources about the new framework. Share knowledge constantly in the team and the community.
      </td>
   </tr>
   <tr>
      <td valign="top">New features
      </td>
      <td valign="top">Unknown how the new design will affect the new features.
      </td>
      <td valign="top">Low / Medium
      </td>
      <td valign="top">Check with the team what are the next 2-3 new features that should be implemented and check against the proposed design.
      </td>
   </tr>
</table>

# Execution Plan

We would like to present an overview on the possible [milestones](#milestones) and the order of work to be performed.

Please note that in internal discussions we concluded that we would like to start implementing the new REST side of Document API only once we have selected the [new framework our microservices](https://github.com/stargate/stargate/discussions/1526) will be based on. We believe that this will be of great benefit and saves us a lot of time. In addition, the new REST side will be started off from a clean state, without any copy/paste of the existing code. We think this is needed in order to achieve above requirements and notice where our existing implementation is not good enough. We also believe this is the only way to migrate everything to a purely reactive execution, as well as to close all testing gaps.

## Milestones

<table>
   <tr>
      <td><strong>#</strong>
      </td>
      <td><strong>Milestone</strong>
      </td>
      <td><strong>Description</strong>
      </td>
      <td><strong>SG V2 Milestone</strong>
      </td>
   </tr>
   <tr>
      <td valign="top">M1
      </td>
      <td valign="top">Finish current refactor
      </td>
      <td valign="top">
         Currently we are refactoring the write path of the documents API. This refactoring will result in unified & cleaner code, better tests and will remove a lot of the old code. Finalizing this refactoring will make further milestones to be easier.
         <p>
            <em>Note that this milestone will be considered as done once we have the TTL feature merged (see <a href="https://github.com/stargate/stargate/pull/1670">stargate/stargate#1670</a>).</em>
      </td>
      <td valign="top">-
      </td>
   </tr>
   <tr>
      <td valign="top">M2
      </td>
      <td valign="top">Finalize and design gRPC API
      </td>
      <td valign="top">Choose a proposal from this document and design proto files for it. If proposal I is chosen, this milestone should also include investigation of other document databases in order to collect best practices.
      </td>
      <td valign="top">M2
      </td>
   </tr>
   <tr>
      <td valign="top">M3
      </td>
      <td valign="top">Implement gRPC server side
      </td>
      <td valign="top">
         Do needed changes in the Bridge and coordinator in order to implement the needed gRPC API changes. 
         <p>
            <em>Note that during this milestone, the existing Document API exposed in the coordinator (v1) should not be modified.</em>
      </td>
      <td valign="top">M2
      </td>
   </tr>
   <tr>
      <td valign="top">M4
      </td>
      <td valign="top">Bootstrap the REST project
      </td>
      <td valign="top">
         Perform bootstrapping of the REST project in order to properly set up all aspects needed to be implemented as per <a href="#rest-service-responsibility">REST Service Responsibility</a>.
      </td>
      <td valign="top">M2/M3
      </td>
   </tr>
   <tr>
      <td valign="top">M5a
      </td>
      <td valign="top">Collection handling on REST
      </td>
      <td valign="top">Implement operations for a collection handling (create, delete) on the REST side.
      </td>
      <td valign="top">M2/M3
      </td>
   </tr>
   <tr>
      <td valign="top">M5b
      </td>
      <td valign="top">Document writes on REST
      </td>
      <td valign="top">Implement write, update, patch and delete operations on the REST side.
      </td>
      <td valign="top">M2/M3
      </td>
   </tr>
   <tr>
      <td valign="top">M5c
      </td>
      <td valign="top">Document reads on REST
      </td>
      <td valign="top">Implement read and search operations on the REST side.
      </td>
      <td valign="top">M2/M3
      </td>
   </tr>
   <tr>
      <td valign="top">M6
      </td>
      <td valign="top">Production ready
      </td>
      <td valign="top">Implement anything needed for making the REST side a production-ready application. Perform any additional tests or performance analysis. 
      </td>
      <td valign="top">M2/M3
      </td>
   </tr>
</table>

_* Please note that milestones don’t have to be done sequentially. It’s clear that some milestones can be done in parallel and that work on some milestones can start before another one which is higher in the list. Each milestone should define a set of tickets, and working on tickets from different milestones at the same time is allowed. However, it’s expected that finalization dates of milestones come in the above order._

## Issues

All issues related to the Document API V2 are marked with the `stargate-v2` and `documents API` [labels](https://github.com/stargate/stargate/issues?q=is%3Aopen+is%3Aissue+label%3Astargate-v2+label%3A%22documents+API%22).

**Comment by clun** [comment]
Thank you for the specifications. Here are some ideas after reading... feel free to comment.

**Architecture**
- At the coordinator node level could we disable the bridge totally (evntually having a different build). Having `no-data nodes` has value even - if we do not use neither `Http` nor `gRPC` apis.

- With so many components, (containers) do you ambition to create a dedicated `K8s` operator or helm chart. That would be quite some configuration to setup and then share among components.

- The cardinality `API <-> bridge` is not clear to me. You could instantiate many Document API containers with one to many coordinator nodes. As such at the document API containers there is a load-balancing component to pick one coordinator.

**Authentication**
- Regarding authentication, authorization why not reusing a standard like CAS, Oauth2 or OpenId Connect. Then user definitions and claims could be externalized to a dedicated third-party solution (keyforge, okta...). Stargate could propose a default authorization server.

**Rest API:**

The current specification details more architecture design than the API interfaces themselves. I understand that very little will change at first. Still moving from `Dropwizard` to `Quarkus` would required quite some code change (swagger->openApi, JaxRS is unchanged), as such i would propose some ideas.

* Some metadata that you want to use in CQL are not reproduced in the Rest API level. I think it could a good addition and allow then as headers params: `consistency level`, `write time` + operations on list.
* Could we have a low-level endpoints where you pass your CQL request as a String (externalizing parameters) `/raw-cql` to reduce serializations concerns. You could also here considering reintroducing the prepared statements.

**Document API**
- At the coordinator node level could we disabled the extension regarding the document API ? Again here the idea would be to start only what you need.

**Reply by ivansenic** [reply]
@clun Hi, thanks for opening the discussions.. I'll try to respond to all items..

**Architecture**

* I can see this as a possibility.. However, this is more Stargate V2 discussion item, then Document API.. But yea, I can see this option possible, no reason not to be..
* Yes, I think @jeffreyscarpenter took the lead on the issue, not sure if Helm is preferred over operator or vice versa..
* The cardinality is many to many, a single Docs API instance can talk to many coordinators and they can be scaled horizontally as well..

**Authentication**

* What we do have at the moment on the Document API should be a so called header-based authentication. However, I already pointed out the possibility to introduce standardized authentication mechanism.. This is definitely on our radar, and something that we should go into rather sooner than later..

**Rest API**

This is most likely best to be answered by @jeffreyscarpenter and @tatu-at-datastax.

**Document API**

* Again I see no reason we could not provide the switch.

**Comment by jeffreyscarpenter** [comment]
On the helm vs. operator discussion. Historically we have had simple docker compose configuration examples under Stargate, and left the Kubernetes deployment details to K8ssandra. (For v2 the docker compose examples are in the `stargate` repo instead of the separate `docker-images` repo which we'll likely deprecate when v2 is complete.) I would be inclined to continue with this approach so that we avoid duplicating K8ssandra. We will need to coordinate with K8ssandra project on deployment of Stargate v2 API services.

**Comment by jeffreyscarpenter** [comment]
> At the coordinator node level could we disabled the extension regarding the document API ? Again here the idea would be to start only what you need.

Since the "extension" is just one additional method on the bridge, I don't think there's a lot of value in that. I suppose we could have an option to turn the entire bridge on or off, but that would imply you weren't using any of the API services.
